OpenSim/Common/DataTable.h

```diff
@@ -0,0 +1,274 @@
+    virtual Array<std::string>& updColumnLabels() = 0;
+
+    /* TODO: Read in the units associated with the columns of the DataTable */
+    /* virtual const Array<Units>& getColumnUnits() const = 0;
+    */
+
+    /** Get the underlying type of data held in the DataTable. Can be any 
+        SimTK numerical type: Real, Vec<M>, Mat<M, N>, SpatialVec, ... */
+    virtual const std::type_index& getDataTypeInfo() const = 0;
+
+    /** Dump the DataTable to an output stream */
+    virtual void dumpToStream(std::ostream &out) const = 0;
+
+//=============================================================================
+};  // END of class AbstractDataTable
+//=============================================================================
+
+
+/** Concrete Implementation of the <tt>DataTable<DataType></tt>
+@tparam DataType  Any numerical type: (SimTK::) Real, Vec<M>, Mat<M,N>, ...*/
+template<typename DataType = SimTK::Real>
+class DataTable_ : public AbstractDataTable {
+//=============================================================================
+// METHODS
+//=============================================================================
+public:
+    virtual ~DataTable_(){};
+
+    /** Construct an empty DataTable */
+    DataTable_() : _type(typeid(DataType)) {}
+
+    /** Construct a pre-sized DataTable to be filled in later. If values
+        are accessed before they are assigned they will be NaN. */
+    DataTable_(size_t nrows, size_t ncols, 
+        const DataType& val = DataType(SimTK::NaN)) :
+        _data(int(nrows), int(ncols), val),
+        _type(typeid(DataType)) {}
+
+    /** Convenience construction of a DataTable from a filename with its source
+        type identified by the file extension. E.g. `results.csv`, `states.sto`
+    @param[in] filename         the name of the data file to be accessed
+    */
+    DataTable_(const std::string& filename) :
+        _type(typeid(DataType)) {
+        std::string daId = FileAdapter::findExtension(filename);
+        auto* reader = FileAdapter::createAdapter(daId);
+        if (!reader){
+            throw Exception("DataTable() no adapter to read file '"
+                + filename + "' could be found.");
+        }
+        reader->setFilename(filename);
+        reader->prepareForReading(*this);
+        reader->read();
+        delete reader;
+    }
+
+    DataTable_& operator=(const DataTable_& dt) {
+        _metaData = dt.getMetaData();
+        _data = dt.getAsMatrix() ;
+        _columnLabels = dt.getColumnLabels();
+        return *this;
+    }
+
+    virtual AbstractDataTable* clone() const override {
+        return new DataTable_(*this);
+    }
+
+    const XMLDocument& getMetaData() const override final { return _metaData; }
+    XMLDocument& updMetaData() override final { return _metaData; }
+
+    size_t  getNumRows() const override { return _data.nrow(); }
+    size_t  getNumCols() const override { return _data.ncol(); }
```

Comment:
The documentation should state (for TimeSeries) if time counts as a column, or should provide an additional method "getNumDataCols().